### PR TITLE
Interpolation

### PR DESCRIPTION
**Pull request by garypolzin**
Interpolate values between two known job_counts (straight line model) and input as a step into estimate_job_count process

[Trello ticket](https://trello.com/c/BM6IkXDu/16-interpolation)

### CODE REVIEW

**Review by MarionJHolloway — COMMENTED** [review]
I think this is very nearly complete. There's a few small bits to tidy up in intepolation.py and I think it would be worth creating an integrated test for the whole model to check calculations are as expected.

File: utils/estimate_job_count/models/interpolation.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+from utils.utils import convert_days_to_unix_time
+from utils.estimate_job_count.column_names import (
+    LOCATION_ID,
+    UNIX_TIME,
+    JOB_COUNT,
+    ESTIMATE_JOB_COUNT,
+)
+from pyspark.sql.types import ArrayType, LongType, FloatType
+from utils.prepare_locations_utils.job_calculator.job_calculator import (
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/estimate_job_count/models/interpolation.py:14]: Will do this as a separate ticket as it's used by 32 scripts in total

File: utils/estimate_job_count/models/interpolation.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+) -> pyspark.sql.DataFrame:
+    date_range_udf = F.udf(date_range, ArrayType(LongType()))
+
+    return df.withColumn(
+        "unix_time",
+        F.explode(date_range_udf(FIRST_SUBMISSION_TIME, LAST_SUBMISSION_TIME)),
+    ).drop(FIRST_SUBMISSION_TIME, LAST_SUBMISSION_TIME)
+
+
+def date_range(
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/estimate_job_count/models/interpolation.py:88]: yep makes sense

File: utils/estimate_job_count/models/interpolation.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+) -> DataFrame:
+    return df.withColumn(
+        new_column_name,
+        F.first(F.col(column_name), ignorenulls=True).over(
+            create_window_for_next_value()
+        ),
+    )
+
+
+def interpolate_values_for_all_dates(df: DataFrame) -> DataFrame:
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/estimate_job_count/models/interpolation.py:155]: very true

File: utils/estimate_job_count/models/interpolation.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            JOB_COUNT,
+            PREVIOUS_JOB_COUNT,
+            NEXT_JOB_COUNT,
+        ),
+    )
+
+    return df.select(LOCATION_ID, UNIX_TIME, INTERPOLATION_MODEL)
+
+
+def interpolation_calculation(
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/estimate_job_count/models/interpolation.py:192]: yep sounds good

File: tests/unit/test_interpolation.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        self.data_for_merging_known_values_df = (
+            generate_data_for_merge_known_values_with_exploded_dates_known_ascwds_df()
+        )
+        self.data_for_calculating_interpolated_values_df = (
+            generate_data_for_interpolating_values_for_all_dates_df()
+        )
+
+        warnings.filterwarnings("ignore", category=ResourceWarning)
+        warnings.filterwarnings("ignore", category=DeprecationWarning)
+
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply tests/unit/test_interpolation.py:41]: done

**Pull request merged**